proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
+    int32_t dimensions;
+    int32_t * dimension_size;
+    void * data;
+  } Array;
+```
+Here, `dimensions` contains information about number of dimensions (necessary
+for multidimensional arrays), `dimension_size` is an array of size `dimensions`
+containing sizes for each dimensions, and `data` is continuous in memory place
+keeping scalar data.
+
+This representation is closer to `std::vector<T>` rather then `T[]`. Then,
+interoperability between C++ and carbon is limited.
+
+One-dimensional and nested arrays have `dimensions = 1`.
+
+### Declaration
+
+During declaration of arrays, it is more importantly to see type of variable.
+It is the aim of the below presented alternatives of declaration.
+
+Current realization:
+```carbon
+  var array_1D: [i32; 4];
+  var array_2D: [i32; 4, 10]; // is not supported yet
+  var array_nested: [[i32; 10]; 4];
+```
+In the case of nested array, definition of them a bit misunderstood due to
+following indexing is inversed.
+
+Suggested in [#1787](https://github.com/carbon-language/carbon-lang/pull/1787):
+```carbon
+  var array_1D: [4; i32];
+  var array_2D: [4, 10; i32];
+  var array_nested: [4; [10; i32]];
+```
+This definition is deprived of a defect of previous definition, but the type is
+the last one.
+
+Alternatives:
+```carbon
+  var array_1D[4]: i32;
+  var array_2D[4, 10]: i32;
+  var array_nested[4][10]: i32;
+```
+In this alternative, `auto` will work only for types, not for dimensions of
+array.
+
+```carbon
+  var array_1D: i32[4];
+  var array_2D: i32[4, 10];
+  var array_nested: i32[4][10];
+```
+This definition requires contextual information since `i32[4]` is the same as
+`array_1D[1]` so it is necessary to separate type and variable.
+
+Also, it is possible to declare array in function with runtime defined size:
+```carbon
+  fn f(n: i32) -> i32 {
+    var arr: [n; i32];
+    return 0;
+  }
+```
+
+Only of size is defined in compile-time and it is less than
+`MAX_STACK_SIZE_ARRAY`, array is located in stack memory. Other definitions are
+located in heap memory. It is especially important for the last example.
+
+### Indexing
+
+Indexing possible via integer values like in C style:
+```carbon
+  var array: [4; i32];
+  var elem: auto = array[1];
+```
+and also via array:
+```carbon
+  var array_ind: [3; i32] = (0, 2, 4);
+  var array: [5; i32] = (4, 3, 2, 1, 0);
+  var arrav: auto = array[array_ind];
+  // arrav == (array[0], array[2], array[4]) == (4, 2, 0)
+```
+Array indexing is useful if slices generate arrays.
+
+Indexes are also possible on left-side expression:
+```carbon
+  var array: [6; i32] = (5, 4, 3, 2, 1, 0);
+  array[(0,1,2)] = array[(3,4,5)]; // array = (2, 1, 0, 2, 1, 0)
+```
+It requires the same sizes on the left and right sides.
+
+## Rationale
```

Comment:
It would be useful to say something about C++ compatibility and interop here, especially since the representation chosen isn't the same as in C++. Is it OK that Carbon's array representation would be different?

Reply:
> Is it OK that Carbon's array representation would be different?

I think that it is OK. The point of usage `void * data` is for simply interoping between C++ and Carbon.

Reply:
How does that work with ownership? Could you copy a c++ array into carbon and then destroy it? What would happen there?